### ISSUE TITLE
Computing elements of an fp-group of order 200 running out of memory

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#5671]
Consider this:
```
gap> F:=FreeGroup(2);; f1:=F.1;; f2:=F.2;;
gap> G := F / [ f2^100, f1^2, f2*f1*f2^-99*f1^-1 ];;
gap> Size(G);  # <- this is crucial!!!
200
```
But then:
```
gap> Elements(G);;
Error, the coset enumeration has defined more than 4096000 cosets
...
gap> ct:=CosetTable(G, TrivialSubgroup(G));; time;
Error, the coset enumeration has defined more than 4096000 cosets
...
```
It all works if I remove the call to `Size` !

Luckily this is easily handled by ACE:
```
gap> LoadPackage("ace", false);
true
gap> TCENUM:=ACETCENUM;;
gap> Elements(G);; time;
106
gap> ct:=CosetTable(G, TrivialSubgroup(G));; time;
8
```

I totally appreciate that coset enumeration is an art and there are bad examples that can make it necessary to define arbitrarily many cosets (though which those are depends on the enumeration strategy).

But this presentation is not one of those, I'd say? The relator `f2*f1*f2^-99*f1^-1` is ugly but not terribly though. Also, it *works fine* if I do not compute the size first.

Note that first simplifying the presentation also makes it work.
```
gap> H:=SimplifiedFpGroup(G);
<fp group of size 200 on the generators [ f1, f2 ]>
gap> RelatorsOfFpGroup(H);
[ f1^2, (f2*f1)^2, f2^100 ]
gap> Size(H);
200
gap> Elements(H);; time;
91
```

Final note: it worked in GAP 4.4.12, but failed in GAP 4.5.

**Comment by james-d-mitchell** [comment]
Again, I know this isn't a solution (or perhaps very helpful) but:

~~~
gap> F:=FreeGroup(2);; f1:=F.1;; f2:=F.2;;
gap> G := F / [ f2^100, f1^2, f2*f1*f2^-99*f1^-1 ];;
gap> Size(G);
200
gap> time;
2830
gap> G := F / [ f2^100, f1^2, f2*f1*f2^-99*f1^-1 ];;
gap> S := AsMonoid(IsFpMonoid, G);
<fp monoid with 4 generators and 7 relations of length 216>
gap> Size(S);
200
gap> time;
6
gap> AsList(S);
[ <identity ...>, f1, f2, f2^-1, f1*f2, f1*f2^-1, f2^2, f2^-1^2, f1*f2^2, f1*f2^-1^2, f2^3,
  f2^-1^3, f1*f2^3, f1*f2^-1^3, f2^4, f2^-1^4, f1*f2^4, f1*f2^-1^4, f2^5, f2^-1^5, f1*f2^5,
  f1*f2^-1^5, f2^6, f2^-1^6, f1*f2^6, f1*f2^-1^6, f2^7, f2^-1^7, f1*f2^7, f1*f2^-1^7, f2^8,
  f2^-1^8, f1*f2^8, f1*f2^-1^8, f2^9, f2^-1^9, f1*f2^9, f1*f2^-1^9, f2^10, f2^-1^10, f1*f2^10,
  f1*f2^-1^10, f2^11, f2^-1^11, f1*f2^11, f1*f2^-1^11, f2^12, f2^-1^12, f1*f2^12, f1*f2^-1^12,
  f2^13, f2^-1^13, f1*f2^13, f1*f2^-1^13, f2^14, f2^-1^14, f1*f2^14, f1*f2^-1^14, f2^15,
  f2^-1^15, f1*f2^15, f1*f2^-1^15, f2^16, f2^-1^16, f1*f2^16, f1*f2^-1^16, f2^17, f2^-1^17,
  f1*f2^17, f1*f2^-1^17, f2^18, f2^-1^18, f1*f2^18, f1*f2^-1^18, f2^19, f2^-1^19, f1*f2^19,
  f1*f2^-1^19, f2^20, f2^-1^20, f1*f2^20, f1*f2^-1^20, f2^21, f2^-1^21, f1*f2^21, f1*f2^-1^21,
  f2^22, f2^-1^22, f1*f2^22, f1*f2^-1^22, f2^23, f2^-1^23, f1*f2^23, f1*f2^-1^23, f2^24,
  f2^-1^24, f1*f2^24, f1*f2^-1^24, f2^25, f2^-1^25, f1*f2^25, f1*f2^-1^25, f2^26, f2^-1^26,
  f1*f2^26, f1*f2^-1^26, f2^27, f2^-1^27, f1*f2^27, f1*f2^-1^27, f2^28, f2^-1^28, f1*f2^28,
  f1*f2^-1^28, f2^29, f2^-1^29, f1*f2^29, f1*f2^-1^29, f2^30, f2^-1^30, f1*f2^30, f1*f2^-1^30,
  f2^31, f2^-1^31, f1*f2^31, f1*f2^-1^31, f2^32, f2^-1^32, f1*f2^32, f1*f2^-1^32, f2^33,
  f2^-1^33, f1*f2^33, f1*f2^-1^33, f2^34, f2^-1^34, f1*f2^34, f1*f2^-1^34, f2^35, f2^-1^35,
  f1*f2^35, f1*f2^-1^35, f2^36, f2^-1^36, f1*f2^36, f1*f2^-1^36, f2^37, f2^-1^37, f1*f2^37,
  f1*f2^-1^37, f2^38, f2^-1^38, f1*f2^38, f1*f2^-1^38, f2^39, f2^-1^39, f1*f2^39, f1*f2^-1^39,
  f2^40, f2^-1^40, f1*f2^40, f1*f2^-1^40, f2^41, f2^-1^41, f1*f2^41, f1*f2^-1^41, f2^42,
  f2^-1^42, f1*f2^42, f1*f2^-1^42, f2^43, f2^-1^43, f1*f2^43, f1*f2^-1^43, f2^44, f2^-1^44,
  f1*f2^44, f1*f2^-1^44, f2^45, f2^-1^45, f1*f2^45, f1*f2^-1^45, f2^46, f2^-1^46, f1*f2^46,
  f1*f2^-1^46, f2^47, f2^-1^47, f1*f2^47, f1*f2^-1^47, f2^48, f2^-1^48, f1*f2^48, f1*f2^-1^48,
  f2^49, f2^-1^49, f1*f2^49, f1*f2^-1^49, f2^50, f1*f2^50 ]
gap> time;
44
~~~

**Comment by dimpase** [comment]
So what does `Size()` do here? This looks like a bug, related  to something stateful being set to something by this call, invalidating future computations with this presentation.

PS. Sorry to nag people here, but how about a new release of ACE? https://github.com/gap-packages/ace/pull/36

**Comment by ChrisJefferson** [comment]
@dimpase : In general, it is very, very common for GAP to have different functions for when it knows the size of a group (look for `HasSize`). Usually these algorithms are more efficient, but it isn't a completely perfect game, and sometimes (like here) the algorithm ends up being worse. Of course, someone could look into exactly why it's worse in this case.

**Comment by hulpke** [comment]
I think the expectations for working with finitely presented groups really are too high, and I am doubtful that it is worth (at least my) time to try to conceive workarounds to make it go smoothly. The obvious answer is to convert to permutation or PC representation rather than try to force the FP methods to handle such situations. Anyhow, a few remarks to this situation:

- The TC in GAP uses the Felsch strategy and therefore works rather badly with long relators (and x^100 is a long relator). ACE implements other strategies and thus is better. I invite anyone interested to update the GAP TC (which is some of the oldest code in the system).
- The `Size` command does a coset enumeration by a subgroup (and then rewrites to this cyclic subgroup), while the elements enumeration does a coset enumeration by the trivial subgroup. If the subgroup is generated by f.2 (which will be one of the easy attempts), this defuses the f.2^100 issue, while it really hits the enumeration by the trivial subgroup.
- The difference depending whether the `Size` (and thus `IsFinite`) is known is due to a special method in line 5509 of `grpfp.gi` that gets triggered if the group is known to be finite. In many cases this approach is more efficient than the generic (Dimino algorithm) code that is used otherwise -- here it leads to the problematic coset enumeration. It is not implausible that this method got introduced in 4.5, thus the difference. Maybe this method is obsolete. One could try instead `Orbit(G,One(G),OnRight)`, or maybe even (if the group is smaller than 10000, say) call `SetReducedMultiplication(G)` beforehand. I welcome anyone interested to experiment what would be a better approach.

**Comment by james-d-mitchell** [comment]
Thanks for the comments @hulpke, but line 5509 of `grpfp.gi` is a blank line in the `master` branch between to method installations, which version of GAP does this line number refer to?

**Comment by james-d-mitchell** [comment]
Oh wait is the method you're referring to, this:

~~~
InstallMethod( Enumerator,"fp gp.", true,[IsSubgroupFpGroup and IsFinite],0,
  G->RightTransversal(G,TrivialSubgroup(G)));
~~~

**Comment by james-d-mitchell** [comment]
Wouldn't one possible solution be to just store the coset table produced by

~~~
    T := NEWTC_CosetEnumerator( FreeGeneratorsOfFpGroup(G),
          RelatorsOfFpGroup(G),GeneratorsOfGroup(H),true,false:
            cyclic:=true,limit:=1+max );
~~~

(in the `Size` method for finitely presented groups) in the group, and then use this in subsequent computations for the group? Surely if the coset enumeration terminates when computing the size of `G`, then there's no point in rerunning this computation (or performing another even slower computation) when the coset table is available? Maybe it is stored in the group object already, but not obviously so:

~~~
gap> KnownAttributesOfObject(G);
[ "Size", "Representative", "OneImmutable", "GeneratorsOfMagmaWithInverses",
  "MultiplicativeNeutralElement", "AbelianInvariants", "FreeGeneratorsOfFpGroup",
  "RelatorsOfFpGroup", "FreeGroupOfFpGroup" ]
gap> NamesOfComponents(G);
[ "Size", "Representative", "OneImmutable", "GeneratorsOfMagmaWithInverses", "AbelianInvariants",
  "FreeGeneratorsOfFpGroup", "RelatorsOfFpGroup", "FreeGroupOfFpGroup", "randomrange",
  "randlim", "randdist" ]
~~~

**Comment by hulpke** [comment]
> Oh wait is the method you're referring to, this:
> 
> ```
> InstallMethod( Enumerator,"fp gp.", true,[IsSubgroupFpGroup and IsFinite],0,
>   G->RightTransversal(G,TrivialSubgroup(G)));

Yes, that is the method. IIRC, it is there to handle the fallback case, of being able to get a faithful permutation representation only as right action on elements.

**Comment by hulpke** [comment]
> Wouldn't one possible solution be to just store the coset table produced 
> (in the `Size` method for finitely presented groups) in the group, and then use this in subsequent computations for the > > group? Surely if the coset enumeration terminates when computing the size of `G`

But what is enumerated is not the cosets of the trivial subgroup but the ones of a cyclic subgroup (which in this example likely is the one of order 100). Then the presentation is rewritten to this cyclic subgroup, and the subgroups order is determined. The coset table thus cannot be used directly to get all elements, but only to induce a representation from the subgroup.

I am pretty sure that the issue here is not a naive oversight, but complications of allowing for various methods (faithful representation as quotient, confluent rewriting system, action on elements) to be used to perform calculations in a finitely presented group. Neither can be guaranteed to work, so all are tried a bit (or as they are found). Fp groups *are* difficult.

**Comment by fingolfin** [comment]
This is resolved in GAP 4.13.1 thanks to PR #5677
